Matroska unleashed

Starwer opened the pull request:
_Although the former Matroska support provided an excellent base for developping it, it was quite limited, and did not enable to save anything.
This branch adds the full support of Matroska (mkv/mka) Tags to TagLib#, plus the support of Nunit and debug tests in Visual Studio 2015 (Windows)._ 

**Fixes and Features in this release include:**
  * Adapted Nunit tests to compile and run on Visual Studio (Windows). 
  * Added debug project to help development
  * Now support all Tags features in Matroska, tag hierarchy, nested SimpleTag, Track/Attachment tagging...
  * Comply to the Matroska standard tags, map the TagLib# Tag properties to the right standard tags;
  * Added support of Attachments (and pictures);
  * Enable to save Matroska tags and attachments;
  * Support of Matroska-specific features (which do not map to the standard TagLib# Tag);
  * Added Matroska examples (mkv video) and Nunit tests;
  * Added the Filename property to the IPicture structure.
  * Fixed issue #65

decriptor commented:
Thanks for doing all of this work. I'll take a closer look a bit later.
It does look like a change might be causing the test project to not run:
```/usr/lib/mono/xbuild/14.0/bin/Microsoft.Common.targets: error : Cannot copy /home/travis/build/mono/taglib-sharp/debug/App.config to /home/travis/build/mono/taglib-sharp/debug/bin/Debug/debug.exe.config, as the source file doesn't exist.```

rboy1 commented:
This would be a great addition!

rboy1 commented:
I just noticed that you've used convention like null safe operators which only exist in .NET 4.5 it won't compile for older versions. I may be wrong but I thought taglib was supposed to be able to compile with .NET 2.0

rboy1 commented:
A few safety check you may want to put in @Starwer 
In Genres Set, I don't think there should be a space after ; and there should be check for a null value like:
`Set("GENRE", null, String.Join(";", value ?? new string[] { }));`

Starwer commented:
Thanks for the in-depth review @rboy1 !
I used .NET 4.5 because that is what is defined in the project files: https://github.com/mono/taglib-sharp/blob/master/src/taglib-sharp.csproj
I didn't see any mention anywhere that TagLib# should be tolerant to .NET 2.0... Could anyone confirm that ?
I'll fix the compile error and put the safety on the Genre property as rightly pointed out by @decriptor  and @rboy1. Should I make another PR for that ?
Feel free to give other suggestions. I pretty much appreciate that.

rboy1 commented:
I'm trying to compile it an older version of .net and it's working great if I just replace the ?. null safe operators. Just suggesting that this one change could make it much more compatible.

Also while trying to test it out, I'm getting an exception (invalid RBML id size) when I try to read back the MKV file that had metadata added using the library

> at TagLib.Matroska.EBMLElement..ctor(File _file, UInt64 position) in D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs:line 79

I'm using standard tags, custom tags and also including a picture (cover art) when writing the metadata, however while reading it back I get the exception.

If this helps:
vector.Data contains just one byte `2` and it's translating to a id_length of 5.

rboy1 commented:
Also oddly when I open the MKV file in VLC player I don't see any of the metadata in the file. Will investigate and come back.

Starwer commented:
> Also oddly when I open the MKV file in VLC player I don't see any of the metadata in the file. Will investigate and come back.
I've tried with PotPlayer and it works fine.
It looks like VLC doesn't support the Tag with a TargetTypeValue. This looks like a limitation of VLC. Actually the TagLib# Matroska will most of the time generate a TargetTypeValue. I guess no value mean level 50, but I couldn't find in the specification where it is described.

rboy1 commented:
Will try PotPlayer, meanwhile I'm noticing that if I write to the MKV using TagLib then it always seems to some exception while trying to read it back, last time was EBML element, then I've seen position exception etc.

Here is an examples of my test:
>
                        MKVTag.Title = Title;
                        MKVTag.Set("SUBTITLE", null, SubTitle);
                        MKVTag.Set("DESCRIPTION", null, Description);
                        MKVTag.Set("TVCHANNEL", null, Network);
                        MKVTag.Set("LAW_RATING", null, Rating);
                        MKVTag.Set("ACTOR", null, MediaCredits);
                        MKVTag.Set("GENRE", null, Genres);
                        MKVTag.Set("SEASON", null, Season.ToString());
                        MKVTag.Set("EPISODE", null, Episode.ToString());
                        TagLib.Picture VideoPicture = new TagLib.Picture(BannerFile);
                        MKVTag.Pictures = new TagLib.Picture[] { VideoPicture };

When I do a `NewTagFile.Save();` it takes a very long time to save the file (I noticed if your project settings you has disabled code optimizing, any specific reason?) but the most importantly I get an exception while trying to read back the file:
`TagLib.File TagFile = TagLib.File.Create(videoFileName)`

Any thoughts?

rboy1 commented:
Okay I checked the file using PotPlayer and it only shows the attachment and the Title, none of the custom tags using Set. See image attachment.

![image](https://user-images.githubusercontent.com/3846367/29378435-0f22f7d8-828d-11e7-8e06-3440fefd06c6.png)

Then I also used MKVInfo after running the file through the above code and I don't see any tags except for Title, Attachment and Duration. No custom tags. How are you testing this?

Starwer commented:
Thanks again for all these testing.
That's odd. I tested it using the Nunit test I've created: MkvFormatTest.cs. Did you try to run the Nunit tests ? Do you run it on Windows or Linux ?
And I also test using the 'debug' project. 
Optimization is disabled on purpose in Debug mode, but I don't think it makes a noticeable difference. Fact is I didn't optimize yet the file-writing, so I can understand why this takes long. The writing involves a lot of data insertion, also in the begining of the file. This surely requires optimization, but I intended to do that as a second step. 

Now I'll try to reproduce your test failure... keep you up-to-date.

decriptor commented:
I'm not sure if there is a fixed target .NET framework for taglib#. If all it requires is removing `?.` to be 2.0 compatible then maybe we should do that for now.

I'd actually like to move this to a .NET standard 2.0 library in the near future instead.

decriptor commented:
Also, for any of these issues that come up would you add a nunit test for them. This way we can avoid ever regression in that area. Thanks!

rboy1 commented:
Windows, I'm just running some of my own tests I whipped up like the one I posted above.

decriptor commented:
That is one of the challenges for a project like this. There used to be a submodule of samples, but that has long since disappeared. Not that I'd advocate committing a 700 mb file :)

decriptor commented on tests/fixtures/Helpers.cs:
```diff
@@ -1,11 +1,21 @@
 
 namespace TagLib.Tests
 {
 	public class Debugger
```
I'm not sure I'd call this `Debugger` or at least not with test constants in here.

decriptor commented:
It looks like a bunch of your `Debugger` class changes broke a bunch of the tests.

Starwer commented:
> To get gtk-sharp on windows you just need to install this 

So simple ! I got it 100% pass for the first time on Visual Studio :)
This trick is now documented in the README

Starwer commented:
Thanks @decriptor !
Here it is: the branch is clean, and run on Mono and Visual Studio (Windows).
I still need to fix the Big files problem, but this is only a design issue (no struggle with the flow) :)

decriptor commented:
I'll try and pull this soon to test on mac and linux, but the fact that everything is passing on travis-ci is a very good sign for Linux at the very least :)

rboy1 commented:
Just my thoughts, I see lot of patches submitted by folks, they're great. I completely get the need to have tests for the patches. However if the person doesn't know/have the time/ability to create those tests, leaving those perfectly great patches would be loss for the program. I've send many of the pull request are great but there hasn't been a response from the devs (for whatever reason). Maybe @decriptor you want to institute an alternative mechanism to accept those patches, maybe have a few "trusted" folks review the code and sign off before merging the pulls. Just my thoughts on how to include the broader community while working through limitations on all sides.

rboy1 commented:
@Starwer I have tons to files to test with from 1MB to 20GB. Happy to do so once you've made your patches.

decriptor commented:
@rboy1 The tests are very useful because once those great patches go in the maintainer is responsible for that code not the person that wrote it. So, should anything happen someone else has to fix it. Also, there are so many different standards that I don't have time to be familiar with each one enough to justify the code change or more importantly validate it. It is possible that it only works in that one case and potentially breaks other use cases?

With tests at least we know it passes some set of conditions and when the next developer comes along and fixes something the tests validate that there aren't any regressions.

decriptor commented:
I'd argue that tests are a baseline and having "trusted" individually to review PRs would be very welcomed :)

rboy1 commented:
I agree, that's a solid baseline, no argument on that. I'm trying to address those pull requests pending where for whatever reason the tests aren't available. Would you leave those bugs for lack of tests?

decriptor commented:
@rboy1 the problem is that it is an assumption that those fixes are correct fixes. If someone wanted to validate that they are correct fixes I'm be more inclined to merge them.

Starwer commented:
I'd say that priority should be to integrate the pending patch as long as these don't break the existing regression (and are reviewed). That must be a very frustrating situation for the people who did the changes to not having anything integrated out of that. Unblocking all pending changes would give a strong message that TagLib# is back to life again, and would stimulate more people to contribute.
IMHO, only after that, we can strengthen the policy to get tests covering every change. 

> If someone wanted to validate that they are correct fixes I'm be more inclined to merge them.

Seems like @rboy1  has already reviewed the changes... and has an impressive Test-suite on his side. Could we call that a conclusive review ?

rboy1 commented:
There's one more issue with .NET version compatibility, Tag.cs, line 376

`subtags.Add(list?[0]);`

should be

`
                            if (list != null && list.Count > 0)
                                subtags.Add(list[0]);
`

Assuming you don't want to add a null to the list.

rboy1 commented:
Actually I can reproduce the problem even with small 100MB files or even 20MB files. For some reason if I use the code I posted above, no file coming out of TagLib is showing any readable tags by MkvInfo

rboy1 commented:
There's an issue somewhere in writing tags, even if I don't write custom tags, just do the standard tags:
`
                TagLib.ByteVector.UseBrokenLatin1Behavior = true;
                NewTagFile = TagLib.File.Create(outputFile);
                TagLib.Matroska.Tag MKVTag = NewTagFile.GetTag(TagLib.TagTypes.Matroska) as TagLib.Matroska.Tag;
                        MKVTag.Title = Title;
                        MKVTag.Genres = Genres;
                        MKVTag.Comment = Description;
                NewTagFile.Save();
`

I get an error while trying to read back the tags and even MkvInfo can't find the tags.

`
TagLib.File TagFile = TagLib.File.Create(videoFileName);
`

> System.ArgumentOutOfRangeException was caught
  HResult=-2146233086
  Message=Specified argument was out of the range of valid values.
Parameter name: position
  Source=TagLibSharp
  ParamName=position
  StackTrace:
       at TagLib.Matroska.EBMLElement..ctor(File _file, UInt64 position) in D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs:line 58
       at TagLib.Matroska.File.ReadTargets(EBMLElement element, Tag tag) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 732
       at TagLib.Matroska.File.ReadTag(EBMLElement element) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 707
       at TagLib.Matroska.File.ReadTags(EBMLElement element) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 680
       at TagLib.Matroska.File.ReadWriteSegment(EBMLElement element, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 448

However reading back tags created by other libraries seems to be working fine.

rboy1 commented:
@Starwer  I came across this while reading up on Mkvpropedit, I'm not sure if you're using the following strategy for reading and writing the tags but it can make a huge performance difference and also probably reduce the chances of errors. Maybe you want to include a flag which allows the dev to set the mode they want to use for reading and writing tags.

> Sets the parse mode. The parameter 'mode' can either be 'fast' (which is also the default) or 'full'. The 'fast' mode does not parse the whole file but uses the meta seek elements for locating the required elements of a source file. In 99% of all cases this is enough. But for files that do not contain meta seek elements or which are damaged the user might have to set the 'full' parse mode. A full scan of a file can take a couple of minutes while a fast scan only takes seconds.

Starwer commented:
> There's one more issue with .NET version compatibility, Tag.cs, line 376

Thanks again.
I fixed this locally. In fact, this is really intended to add a null to the list, because when accessing nested SimpleTag, like ARTIST/SORT_WITH, we expect the list to match the list of ARTIST/SORT_WITH, i.e. for each Performers[i], there should be a corresponding PerformersSort[i] (or null if no PerformersSort).

Starwer commented:
@rboy1 : Thanks for the MkvPropEdit trick. I use MkvInfo and MkvToolNix GUI.

rboy1 commented:
@Starwer  some great news and some bad news.

1. The Great news is that with the latest patches the metadata writing is working and TagLib is also able to read it back without an exception, including custom tags.
2. The other good news is that VLC and PotPlayer are able to see the metadata tags

There are 3 bad news:
1. MkvInfo doesn't show the tags and stops at the first cluster (without details) - see image
![image](https://user-images.githubusercontent.com/3846367/29442177-f14a95a6-839d-11e7-9efe-802589d150ae.png)
2. The reading and writing (especially, takes 100x+ compared to reading) takes a VERY VERY VERY long time for large files. Is there a way you can use to seek elements to speed it up like mkvpropinfo?
3. Writing the picture attachment has stopped working (in fact this was the only one that worked earlier, writing and then reading back), now when I attach a picture and try to read it back (see sample code above) TagLib throws an exception:

> TagLib.CorruptFileException was caught
  HResult=-2146233088
  Message=invalid EBML id size
  Source=TagLibSharp
  StackTrace:
       at TagLib.Matroska.EBMLElement..ctor(File _file, UInt64 position) in D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs:line 79
       at TagLib.Matroska.File.ReadWriteSegment(EBMLElement element, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 416
       at TagLib.Matroska.File.ReadWrite(ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 366
       at TagLib.Matroska.File..ctor(IFileAbstraction abstraction, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 170   at TagLib.File.Create(IFileAbstraction abstraction, String mimetype, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\File.cs:line 1475
       at TagLib.File.Create(String path, String mimetype, ReadStyle propertiesStyle) in 

Great work!

PS - I'm still testing it with very large files (7GB files) - the writing time is just killing it!

Starwer commented:
@rboy1: thanks again for the feedback (good and bad).
To me it's only good news, as the outcome is the one I expected. There are some reliability issues in the writing which shades what got previously written. It could be Tags, or Attachments.
I didn't fix yet the performance/reliability issue in the SeekHead writing. Fact is I plan to totally rewrite the Save() part in a more robust, maintainable and efficient way. This involve a better EBML hierarchy representation.
On the other end, I expect that reading is about a second (or two) only, whereas writing is ridiculously long. If not, let me know so that I can indeed optimize it as well.

rboy1 commented:
Happy to hear that you have a handle on it. The read for large files (7GB) isn't a few seconds, more like two minutes (mkvInfo took a few seconds), where as the write took a few hours.

The other news is that when writing to very large files (>5GB) it results in an exception after a few hours and doesn't complete. Here is the stack trace:

> System.OverflowException was caught
  HResult=-2146233066
  Message=Arithmetic operation resulted in an overflow.
  Source=TagLibSharp
  StackTrace:
       at TagLib.Matroska.EBMLElement.set_DataSize(UInt64 value) in D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs:line 312
       at TagLib.Matroska.EBMLElement.ResizeData(Int64 offset) in D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs:line 528
       at TagLib.Matroska.File.ReadWriteSegment(EBMLElement element, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 502
       at TagLib.Matroska.File.ReadWrite(ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 366
       at TagLib.Matroska.File.Save() in D:\TagLibSharp\TagLib\Matroska\File.cs:line 223

Starwer commented:
Ok, this time I have rewritten the whole writing of the Matroska. It's still not a snap, but it has now a workable delay and should be way more reliable.
I still plan to optimize on the reading, but not so much on the writing.
@rboy1: in other words: writing could be tested now ! ;)

rboy1 commented:
@Starwer great work, the tags now appear to be written and read back without exceptions. However some results here. When writing the file and one calls the save method, for a 6GB file it's already been an hour and it's still saving. It ideally shouldn't take more than a minute or few minutes to completing writing the metadata (based on a benchmark from AtomicParsley).

Here are some stats I collected while trying to analyze why it's taking do long to write metadata to a 6GB file

The hard disk is showing about 40% utilization, which means a lot of time is being spent in the code by the CPU and not writing to the disk. Another thing I noticed while writing is that the hard disk is spending 70% reading and only 20% writing. So the code appears to be spending a lot of time reading the file for some reason. Other stats I collected, the average file read request is only 4K per read where as the average file write request is 400K per write.

I'm wondering why there is so much reading happening when writing metadata, is there a way to optimize this, like put the tags at the end of the file instead of inserting them in the middle?
Were you able to implement the mkvprop optimizing of using the meta seek elements to locate the right places and is there a way to reduce amount of reading being done while writing? Also wondering if using a larger read buffer might help if reading is required? Am still curious as to why so much data is being read while writing.

rboy1 commented:
@Starwer I'm also seeing this exception while writing to large MKV files. Have you tried saving to a large file?

> System.OverflowException was caught
  HResult=-2146233066
  Message=Arithmetic operation resulted in an overflow.
  Source=TagLibSharp
  StackTrace:
       at TagLib.Matroska.EBMLreader.set_DataSize(UInt64 value) in D:\TagLibSharp\TagLib\Matroska\EBMLreader.cs:line 326
       at TagLib.Matroska.EBMLreader.ResizeData(Int64 offset) in D:\TagLibSharp\TagLib\Matroska\EBMLreader.cs:line 542
       at TagLib.Matroska.File.ReadWriteSegment(EBMLreader element, ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 525
       at TagLib.Matroska.File.ReadWrite(ReadStyle propertiesStyle) in D:\TagLibSharp\TagLib\Matroska\File.cs:line 366
       at TagLib.Matroska.File.Save() in D:\TagLibSharp\TagLib\Matroska\File.cs:line 223

rboy1 commented:
@Starwer  looks like you're familiar with MKV formats. I found the source code for MKVPropEdit and wondering if you would like to take a look at it to see how they're doing it and maybe get some pointers on how to fix the overflow and the speed issue.

https://github.com/mbunkus/mkvtoolnix/tree/master/src/propedit

mbunkus commented:
Hey, the MKVToolNix author here. One of you contacted me via email about how mkvpropedit works and why it's fast(ish), and I've replied in length, but I think posting it here may be useful to you, too.

> 1. It takes hours to write, we're missing some optimization strategy on how
> you're using MKVPropEdit to write rather quickly

What mkvpropedit does is somewhat involved. Matroska is a rather flexible format, and that has its price: updating existing data without re-writing everything to a new file is a complex task.

First, let's take a look at reading a file. For this mkvpropedit basically just scans the heads of level 1 elements (`SegmentInfo`, `Tracks`, `Cluster`, `Chapters`, `Attachments`, `Void`…):

* remember current file position
* read EBML ID
* read its size = n
* seek forward n bytes
* rinse, repeat

That's what mkvpropedit calls the "full" parse mode (default is "fast"). The difference with "fast" is that mkvpropedit doesn't read all level 1 elements that way. It only reads until it finds the `MetaSeek` and `Cues` elements and uses the information stored in those elements instead of reading the headers of all level 1 elements. The drawback of the "fast" method is that `Void` elements aren't referenced from `MetaSeek` and `Cues` and will therefore not be known to the updating algorithm — and `Void` elements are a central part of updating an existing file.

Now on to updating the file. This is much more complex than reading the file. I recommend you read the source code of my [`kax_analyzer_c` class](https://github.com/mbunkus/mkvtoolnix/blob/master/src/common/kax_analyzer.cpp). It's one of the few files that I've documented more thoroughly. Use the `update_element` function as the entry point. It's supposed to be called with a new instance of a level 1 element, e.g. `Tracks` (meaning the track headers). It will then use the information gathered while reading the file in order to determine where it writes that element.

The steps it does are:

1. First it determines if the last level 1 element in the file has an unknown size. If so, it'll fix that so that the last level 1 element ends where the file ends. This is required as the analyzer may have to write the Element-To-Update at the end of the file, and in that case the former last level 1 element's size must be known.
2. Next it overwrites all existing instances of the Element-To-Update (e.g. there may be multiple instances of KaxTracks in the file) with `Void` elements.
3. Now it takes a look at all existing `Void` elements and merges all consecutive `Void` elements into a single `Void` element. This is rather common: mkvmerge writes a `Void` element after the `Tracks` element by default. And after step 2 that Tracks element will have been overwritten with a `Void` element, meaning we now have two consecutive `Void` elements.
4. Next step is writing the Element-To-Update. For this the analyzer will use the first `Void` element large enough to contain that element (it'll overwrite the `Void` with the Element-To-Update and create a new, smaller `Void` element directly afterwards covering the remaining space of the `Void` element that has just been overwritten). If no such `Void` element is known, then the element will be written at the end of the file.
5. We now have to fix up any references to the former level 1 elements removed in step 2, meaning we look at all `MetaSeek` elements and remove references to the overwritten elements from them. This means shrinking `MetaSeek` elements, and shrinking in turn means writing new `Void` elements for the space formerly covered by the `MetaSeek` elements.
6. Therefore we now have to merge consecutive `Void` elements again.
7. Our newly-written element should also be referenced from MetaSeek elements, therefore we now add it to them and write the (now bigger) `MetaSeek` element back to the file shrinking any `Void` element located behind it.
8. We finally do one more merge step across `Void` elements.

Sounds complicated already? Well, the devil is in the details, and I highly recommend you read the source of all those functions for additional comments which do mention some of those details. For example: I've mentioned in step 4 above that writing the Element-To-Update to a place currently occupied by a `Void` element requires writing a new, smaller `Void` element directly afterwards covering the remaining space of the just overwritten `Void` element.

Let's take a `Void` element occupying 100 bytes in total (including its EBML ID and its size component) as an example. If we want to write a `Tracks` element there that's 80 bytes in total, then we have to create a new `Void` element occupying 20 bytes total directly after the `Tracks` element.

Simple enough.

However, there's a minimum number of bytes a `Void` element can fit into: two bytes. Why? Because the EBML ID is one byte and the length/size field is at least another byte — meaning you cannot create a `Void` element occupying one byte in total.

So what happens in the example above when the `Tracks` element occupies exactly 99 bytes? A one-byte `Void` element is impossible as we've just learned. So that's out. Leaving space between two elements violates the specs, so we cannot just fill it with 0x00 bytes (it may work with some readers, but that's just because they're flexible and error-resilient — don't rely on it!).

What we can do, however, is make the `Tracks` element slightly larger. Each EBML element has a size/length field, and that field is usually stored with the fewest number of bytes possible. But it is completely OK to store it with more bytes than necessary. Coming back to our example, the length of 99 bytes would normally have to be stored in only one byte: 0xe3 (the first bit indicates that the size/length field is one byte long, the remaining seven bits = 0x63 are the length, decimal: 99). But storing the same length 99 as two bytes with the following sequence is semantically equivalent, allowed by the specs and even encouraged for situations such as the one I'm describing: 0x40 0x63 (the first two bits are 0 1 indicating that the size/length field is two bytes long, the remaining sixteen bits = 0x00 0x63 are again just decimal 99).

So in order to cover that one byte after the `Tracks` we write the 99 bytes long `Tracks` but enlarge its size field by exactly one byte as described above.

And that's not the only pitfall.

The `kax_analyzer_c` class is probably the one single point in the program where I've spent the most amount of time and thought. It's pretty bug-free nowadays, but it took quite some time to get there.

> 2. What we write, we can read back and so can other players like VLC etc but MKVInfo doesn't seem to recognize the tags.

Please note that mkvinfo doesn't scan the whole file by default. Read the following two FAQ entries: ["After using mkvpropedit or the GUI's header or chapter editors all track information has vanished"](https://github.com/mbunkus/mkvtoolnix/wiki/All-tracks-vanished-after-mkvpropedit,-the-header-or-chapter-editors) and ["Understanding mkvinfo's output"](https://github.com/mbunkus/mkvtoolnix/wiki/Understanding-mkvinfo's-output). If they don't explain what you're seeing, then you can [send me a file](https://mkvtoolnix.download/contact.html) and I'll take a look. If you do, please include information about where in the file (byte position) your modification starts and what you expect to see.

mosu

Starwer commented:
Hey ! 
First of all, a big thanks to @rboy1 and @mbunkus to there thorough reviews/explanations ! I'm glad to see how involved and interested people are in this initiative, and this is quite motivating. 
First the good news: 
The reading is now optimized for using the meta-seek sismilar to the @mbunkus's fast mode. However, I don't use the Cues elements. What should I use it for ? The only addition (maybe) is that it detects a meta-seek corruption to fall-back on the "full" mode. Doing that, I'vz achieved a reading of a 9.8 GB file in less than a second. I think I can stop optimizing that part.
The bad news is that this change is not yet available as I have corrupted the wrting part, due to (again) massive restructuration.
For the writing, I now also try to avoid moving data and writing data twice... I try to do it more cleverly as described by @mbunkus by taking advantage of the Voids. That quite a puzzle game indeed... but that's also why it's fun ! :)
@mbunkus: about Step 1, I don't understand... how can an EBML element have an unknown size ? Did I miss something in the spec ? Also, for writing, I plan to fully rely on the fast read to make the map of the Segment top EBMLs... To find the Voids, I'll assume that the Voids will be mapped following the "real" EBML element, i.e. Iùll look after Track to see whether there is an Void hidden there. That way I avoid the reading of the long list of Cues. Does it sounds reasonable to you ?

All in all there is progress, but the Write optimization will take some time to complete...

Starwer commented:
@rboy1: About the bug you reported: I've checked the uint to ulong convertions and fixed these. Let's seen commit what comes out of that... Thanks for the hint !

mbunkus commented:
> @mbunkus: about Step 1, I don't understand... how can an EBML element have an unknown size ? Did I miss something in the spec ?

You did, yes :grin: An element has an "unknown" size if all of the bits of the size/length field's value portion are set to 1 (e.g. for a 1-byte size/length field: 0xff, for a 2-byte size/length field: 0x7f 0xff etc.). This means that the element extends to the end of the file. For example, mkvmerge first writes the segment's size as an unknown size and only updates it once it knows exactly how long the file will be (so you won't find an unknown segment size in files created by mkvmerge unless you forcefully kill mkvmerge in the middle of writing).

Read more about it [in the EBML specification](https://github.com/Matroska-Org/ebml-specification/blob/master/specification.markdown#element-data-size) (third paragraph).

> Also, for writing, I plan to fully rely on the fast read to make the map of the Segment top EBMLs... To find the Voids, I'll assume that the Voids will be mapped following the "real" EBML elements, i.e. I'll look right after Track for example to see whether there is an Void hidden there. That way I avoid the reading of the long list of Cues. Does it sounds reasonable to you ?

Sounds reasonable. Reading the list of Cues itself is usually fast; reading all the clusters fully that those Cues reference is what is slow, though. However, you don't really need the Cues if you just assume that space that's not explained by the content of MetaSeek is opaque and may not be modified.

Starwer commented:
@mbunkus: Thanks for all these valuable tips ! I didn't know this link to EBML Spec, I only used the Matroska spec; https://www.matroska.org/technical/specs/index.html ...
This one really clarifies some blind spots ! :)
I'll handle the Auto-size then.

mbunkus commented:
There's an ongoing effort by the IETF's Cellar working group to standardize EBML, Matroska, FFV1 and FLAC as IETF standards for video archiving purposes. The EBML specs I've linked to is one of the results of that effort. We're currently focusing on the Matroska specs (well, turning the not-so-specy Matroska website into real, proper IETF specifications).

rboy1 commented:
@Starwer those conversions were happening in a couple of places but I'm assuming you've seen that :)

Starwer commented:
> those conversions were happening in a couple of places but I'm assuming you've seen that :)

@rboy1: not sure I see what you mean here... what other couple of places are discussing this matter ?
@mbunkus: I've read the spec, and I still miss a few feature in the TagLib sharp Matroska support. I guess I also have to support the CRC and zero-size (0x00), which are not there yet.

mbunkus commented:
@Starwer CRC support would sure be nice, but not a lot of programs support CRC at all at the moment. For example, no tool in the MKVToolNix package supports them properly yet (which is mostly due to problems in the libraries MKVToolNix uses for Matroska access, `libEBML` and `libMatroska`).

rboy1 commented:
@Starwer I was referring to the ulong to unit issue, in the same file there a few more places where ulong was being typecast to a uint but I'm assuming you caught those also

Starwer commented:
This time, the performance is fixed. I've tried it on a 10 GB file.

- Read < 1 sec
- Write < 1 sec

The Read can take longer in case the SeekHead is absent, in the wrong place or corrupted.
The Write can take longer in case there was not sufficient space at the beginning of the segment to put the SeekHead. The spec says, more or less clearly, that 

> Placing the first Meta Seek Head other than the first position of the Segment would make it a lot less useful So it must be the first element of a Cluster.

And this greatly improve the performances of the subsequent reads. As a result, the SeekHead is always put in the first position in the Segment. 
In those worst cases, we get:

- Read: 36 sec (= time to read all the file);
- Write: 4 min (= time to copy the file);

CRC is tolerated but not really supported, I've added the +1 Size optimization, and support of the 0x00 and 0xFF magic numbers for Data-size.
@rboy1: the bugs in conversion should have disappeared, together with the code that contained these...

I think now we are good :)

rboy1 commented:
@Starwer im seeing a compilation error in 3 places EBMLElement.cs and File.cs where you are calling Insert

> Error	5	Argument 1: cannot convert from 'long' to 'TagLib.ByteVector'	D:\TagLibSharp\TagLib\Matroska\EBMLElement.cs	424	29	TagLibSharp

> file.Insert(size - reserved, position + reserved);

rboy1 commented:
I think you may need to do new ByteVector((int)(size - reserved)), not sure how it compiled for you. Can you verify that the value being type casted from long to int won't cause an overflow in these Inserts?

rboy1 commented:
Apart from this, it's working like a charm, it's super fast both reading and writing. Will do more tests but so far it's awesome. You guys are great!!

Final point to note whatever metadata is written can't be read by MKVInfo, I don't know if it's a compatibility issue or not.

rboy1 commented:
Brilliant working like a charm passed all my tests including a 18GB file took less than a second to read and write the tags. Unicode worked perfectly.

Just confirm my changes above for the build errors.

Only point to note is that MkvInfo isnt' able to read the metadata. @Starwer  and @mbunkus  if there are any thoughts on this one.

mbunkus commented:
If the data is placed at the end, then mkvinfo won't show it by default. Please read [this FAQ entry](https://github.com/mbunkus/mkvtoolnix/wiki/Understanding-mkvinfo's-output).

Starwer commented:
> Only point to note is that MkvInfo isnt' able to read the metadata.

I use Mkvinfo (GUI) to debug my files, and for it to show everything you will have to tick all the options in the menu "Options". This takes way longer to read the file, but this works. This is because, as indicated by @mbunkus, TagLib# will place the edited data at the end if there is not enough space in the Voids or replaced data to insert these earlier. That the trick enabling to taking the Write time from 4 min to less than a second. If you use command line, you should use the -v (verbose mode). 

> Error 5 Argument 1: cannot convert from 'long' to 'TagLib.ByteVector' 

@rboy1 : That's odd. it compiles fine (not even a warning) on Travis and on my Visual Studio 2015. It looks like you have an outdated taglib-sharp/src/TagLib/File.cs file, as it should see the new overloaded function Insert(long, long) coming from my branch. This function enable to avoid writing twice on the file when inserting a new Void, and is then way cleaner. The ByteVector((int)(size - reserved)) must be avoided, as it is most costly in memory and will write twice the inserted area on the file. Maybe check whether you have local change on that file obstructing the incoming updates to apply.

rboy1 commented:
Okay just to be clean, I downloaded your latest branch here:
https://github.com/Starwer/taglib-sharp/tree/matroska_unleashed

When I compile it I get the following error:
> Error	23	Identifier expected; 'static' is a keyword	D:\TagLibSharp\TagLib\Ogg\Codecs\Vorbis.cs	25	7	TagLibSharp

The issue is with this definition
>using static TagLib.Ogg.Codec;

Are you compiling from the same branch?

Starwer commented:
> using static TagLib.Ogg.Codec;

@rboy1 : odd that it shows up only now at your place... I'll check that. I've added this line 15 days ago to avoid a warning, but I guess it was a wrong move. But I'll check on my setup why this doesn't show up... (rebuild all). And yes, it is ok to remove it, no doubt about that.

decriptor commented:
@Starwer it looks like you added a few `using static` which is a c# 6 feature, but we are targeting 4.5 (c# 5). That might be the reason @rboy1 is hitting the compile error. It looks like that are probably not needed so we could just remove them?

decriptor commented:
https://github.com/mono/taglib-sharp/pull/68/commits/96ef076f68f7b504e882d53a74f44d540ecf37b7#diff-53915bec962bb0f08a54fa7ec1d60ccfR26

decriptor commented:
@Starwer a couple things real fast. Mixing in unrelated changes like fixing doc stuff makes the PR much harder to parse. They are good changes, but in future if you can put them in a separate PR that'd be great.
Also, You have replaced `tabs` with `spaces`. Would you be willing to go back and change them back to `tabs`? That way we keep the format consistent. Thinking about this we should probably come up with an .editorconfig file.

rboy1 commented:
I'm targeting a 4.0 client profile build so that it's compatible with XP and many older libraries and build environments.

Starwer commented:
@decriptor : I acknowledge that... I have been a bit over-enthusiastic about on getting rid of the warnings. Sorry about this. Next time, I'll take it easy on the change... I've already hold some ideas I have in mind until this Matroska branch gets integrated (see #69).
About the tabs: in fact, I was not aware that Visual Studio were changing the indenting scheme. There is already an .editorconfig file, but seems like Visual Studio doesn't give a thing about it. Default config for Visual Studio in: use spaces. I apologize for that, I personally also hate when we mix up two kinds of indentations. So 4-spaces hard tab it should be now.
As a result, I've done a massive change/replace of leading spaces to tab in the *.cs files. At least this should make it consistent (I missed pretty much my beloved grep/sed combo in Windows for that).

@rboy1: I've removed the "using static", and in addition I've fixed a few errors that were popping up when switching to .NET 2.0. I've now switched the project to .NET 4.0 (it was .NET 4.5), to make it a bit closer to the target. Still, it didn't show any problem in .NET 4.0 on the "using static" constructs...

decriptor commented:
@Starwer no worries. I've been guilty of this too. As for the editorconfig I think VS2017 supports it, but before that you need an extension?
btw, one more little formatting thing. With methods it should be tabbed up to the beginning of the method signature and then spaces until the parameter lines up with the first one.
ie, https://github.com/mono/taglib-sharp/pull/68/files#diff-6c4460a18efad3806d33dbb1ee8ed240R415

@rboy1 if you want to review this once more I think this is pretty much ready to merge.

rboy1 commented:
I haven't been though the tab changes but did take at the .NET compatibility changes and they look good to go. If we're targetting .NET 4.0 then even those changes (barring the static using) aren't required.

@decriptor your call on whether you want to target .NET 2.0 or 4.0 - no big difference, just no linq and few other things.

rboy1 commented:
Doesn't .NET 2.0 support toArray() and .Last()? I noticed a few places that was replaced in the last clean up.

decriptor commented:
I'm not sure what the user base is, so not sure on the version of .NET. I'd be happy to support C# 6.0 or at least 5.0. However, I'd still push to not use Linq at all from a perf POV.

rboy1 commented:
In that case I would highly recommend staying with 4.0 as it makes it very compatible. Currently the code checked in fully compliant with 4.0 standard and compiling.

Starwer commented:
@descriptor: hum... the function signature isn't something easy to solve in a regular expression point of view... I'll try to roll-back to version before the 4-space change and try a more sophisticated regexp...

@rboy1: ToArray(), Last() and AddRange() didn't fly if I switched to .NET 2.0. I think it has to do with that these are extension function, from Generic or even Linq... anyway, this counted for 10 locations where these were used, so no big deal.

Starwer commented:
@decriptor : This time, I have redone the spaces to tabs indentation conversion, but I have preserved the tabs+spaces constructs (used in the function declaration).
I guess it should be perfect now ? :)

decriptor commented:
@starwer: Thanks!
@rboy1 if you want to give a final sign-off on this I think I'm ok to merge this now.

rboy1 commented:
There are way too many changes in the last 2 commits but I'll take Starwer at his word that they're all related to indentation. Looks good. Great work guys, thanks!

decriptor commented:
@rboy1 thanks for all of the testing on this! If you are willing to test other PRs and clear them I'd be happy to seriously consider merging them.

rboy1 commented:
Happy to do it, I've tested quite a few of them.

decriptor commented:
@rboy1 if you want to comment on them that'd be great. If you'd be willing to create a unit test for them that'd be amazing! Or if you would be willing to more or less come up with what a reasonable test would test I'd be willing to write the test. Part of the reason I haven't done that is because I don't know the specs or have the time to really dig into it :(

Starwer commented:
Thanks to all of you @descriptor, @rboy1 and @mbunkus ! You were a great support on this topic ! Looking forward to working with you again on #69 ! :)
TagLib# rocks !
